### ISSUE TITLE
torch.log has unexpected gradient over negative domain

### ISSUE THREAD

**Issue by louixp** [pytorch/pytorch#76516]
### 🐛 Describe the bug

When the input is real, the log function and its derivative are undefined over the negative domain. In torch, the log function is undefined, but its derivative evaluates to 1/x over the negative domain. I noticed this when my loss has become undefined but the gradients are still defined. I'm not sure if this is a feature or a bug.

```python3
import torch

t_real = torch.tensor([-1.0], requires_grad=True)
l = torch.log(t_real) # This is NaN, as expected.
l.backward()
print(t_real.grad) # This is -1, which is not expected.
```

### Versions

PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3.1 (x86_64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.3)
CMake version: version 3.19.1
Libc version: N/A

Python version: 3.10.0 (v3.10.0:b494f5935c, Oct  4 2021, 14:59:20) [Clang 12.0.5 (clang-1205.0.22.11)] (64-bit runtime)
Python platform: macOS-12.3.1-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by ngimel** [comment]
Thanks for reporting!

**Comment by Lezcano** [comment]
Note that when the input is not in the domain of the function, the rules we follow at PyTorch is that the derivative is not defined, so we are allowed to return anything. In particular, we may return the most efficient value for this, which is that given by the formula  `grad / input`, which is what's happening here. Interestingly enough, this is the value that you would get for the gradients if the inputs were complex (note that in this case the log is well-defined away from zero).

For reference, these are the rules we follow:
>To try and reduce the impact of this limitation, we define the gradients of the elementary operations by applying the following rules in order:
>
>- If the function is not defined (:math:`\sqrt(-1)` or :math:`\log(-1)` for example) then this is undefined. Most function will return :math:`nan`, but for performance reasons, some functions will return not-nan values (:math:`\log(-1)` for example).
>- If a gradient exist at the current point, return it.
>- If the function is convex, return one of it's subgradients at the current point, if more than one exists, return the subgradient with minimum norm (as it is always a descent direction, see "Convex Optimization Algorithms" by Dimitri Bertsekas, Proposition 4.3.1 for a proof).
>- Define the gradient at the current point by continuity (note that :math:`inf` is possible here). If multiple values are possible, pick one arbitrarily.


We could "fix" this but that'd be at the expense of making the formula either slower (calling more kernels) or implementing it via TI, and effectively moving the problem to the second derivative. I don't have a strong opinion for one or the other. @albanD thoughts?

**Comment by albanD** [comment]
I think the rules as specified above are good. And I don't see any reason why we should be careful about the gradient being returned for a function that is evaluated outside of its domain.
Note that while in this case it could be relatively simple to "check" if we're in the domain or not, there are cases in linalg functions where that would be very expensive and we don't want to do it.
